sinkhole-core/src/elgamal/mod.rs

```diff
@@ -0,0 +1,153 @@
+use crate::traits;
+use traits::core::Storage as storage_trait;
+
+use errors::errors::ServerError;
+use rand_core::OsRng;
+use std::cell::RefCell;
+
+use curve25519_dalek::constants::RISTRETTO_BASEPOINT_TABLE;
+use curve25519_dalek::scalar::Scalar;
+use elgamal_ristretto::ciphertext::Ciphertext;
+use elgamal_ristretto::private::SecretKey;
+use elgamal_ristretto::public::PublicKey;
+
+struct Storage {
+    secret_key: SecretKey,
+    size: usize,
+    store: RefCell<Vec<Scalar>>,
+}
+
+impl Storage {
+    fn new(sk: SecretKey, store: Vec<Scalar>) -> Self {
+        return Storage {
+            secret_key: sk,
+            size: store.len(),
+            store: RefCell::new(store), // mutable interior of Cell
+        };
+    }
+
+    fn new_empty(sk: SecretKey, size: usize) -> Self {
+        return Storage {
+            secret_key: sk,
+            size: size,
+            store: RefCell::new(vec![]),
+        };
+    }
+}
+
+impl storage_trait for Storage {
```

Comment:
Because the namespace is so short we could probably just do:

```suggestion
impl traits::core::Storage for Storage {
```